libraries/state_history/trace_converter.cpp

```diff
@@ -30,25 +38,262 @@ void trace_converter::add_transaction(const transaction_trace_ptr& trace, const
+   converter.onblock_trace.reset();
+   return traces;
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const eosio::state_history::augmented_transaction_trace& obj, const Lambda& lambda) {
+   auto& trace = *obj.trace;
+   if (trace.failed_dtrx_trace) {
+      for_each_packed_transaction(
+          eosio::state_history::augmented_transaction_trace{trace.failed_dtrx_trace, obj.packed_trx}, lambda);
+   }
+   bool include_packed_trx = obj.packed_trx && !trace.failed_dtrx_trace;
+   if (include_packed_trx) {
+      lambda(*obj.packed_trx);
+   }
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const std::vector<eosio::state_history::augmented_transaction_trace>& traces,
+                                 const Lambda&                                                         lambda) {
+   for (const auto& v : traces) {
+      for_each_packed_transaction(v, lambda);
+   }
+}
+
+void pack(fc::datastream<char*>& ds, const prunable_data_type&  data, compression_type compression) {
+   if (compression == compression_type::none) {
+      fc::raw::pack(ds, data);
+   }
+   else if (compression == compression_type::zlib) {
+      bytes uncompressed = fc::raw::pack(data);
+      fc::raw::pack(ds, zlib_compress_bytes(uncompressed));
+   }
+}
+
+size_t pack_prunable(bytes& buffer, const packed_transaction& trx,
+                  compression_type compression) {
+
+   const auto& data = trx.get_prunable_data();
+
+   int max_size = data.maximum_pruned_pack_size(compression);
+   uint64_t start_buffer_position = buffer.size();
+   buffer.resize(buffer.size() + max_size);
+   fc::datastream<char*> ds(buffer.data() + start_buffer_position, max_size);
+
+   if (data.prunable_data.contains<prunable_data_type::full_legacy>()) {
+      // convert full_legacy to full before serialization
+      const auto& full_legacy  = data.prunable_data.get<prunable_data_type::full_legacy>();
+      prunable_data_type pd    = {prunable_data_type::full{full_legacy.signatures, full_legacy.context_free_segments}};
+      pack(ds, pd, compression);
+   }
+   else {
+      pack(ds, data, compression);
+   }
+   
+   buffer.resize(start_buffer_position + ds.tellp());
+   return max_size;
+}
+
+
+void unpack_prunable(const char* read_buffer, fc::datastream<const char*>& ds, prunable_data_type& data,
+                     compression_type compression) {
+
+   if (compression == compression_type::none) {
+      fc::raw::unpack(ds, data);
+   }
+   else {
+      fc::unsigned_int len;
+      fc::raw::unpack(ds, len);
+      bytes uncompressed = decompress_buffer(read_buffer + ds.tellp(), len, compression);
+      ds.skip(len);
+
+      fc::datastream<const char*> uncompressed_ds(uncompressed.data(), uncompressed.size());
+      fc::raw::unpack(uncompressed_ds, data);
+   }
+}
+
+bytes pack_prunables(const std::vector<augmented_transaction_trace>& traces, compression_type compression) {
+
+   std::vector<char> out;
+   out.reserve(1024);
```

Comment:
It might be worth getting a reserve amount by looping through traces and adding up the `get_prunable_size()` results.

Reply:
it's hard to reserve the right amount of memory for compression without actually compressing it.